Это вторая часть задания.

Первая часть задания (задания 3.1-3.4) находится в [ноутбуке про линейную регрессию](https://colab.research.google.com/drive/1eX21oBHRAzNr3PimxhMJENF3yeBC2DBe?usp=sharing).

# Предсказание цен на недвижимость

Компании по продаже недвижимости оценивают её стоимость, используя методы машинного обучения. 
Задача оценки цены на недвижимость также может быть нужна для выявления наиболее выгодных предложений, или, например, для выявления аномально низких цен на объект недвижимости.

**Постановка задачи анализа данных** 

Целью данной задачи является прогнозирование стоимости домов в округе Кинг (штат Вашингтон, США) с помощью построения регрессионных моделей и их анализа. Набор данных состоит из цен на дома в округе Кинг, проданных в период с мая 2014 года по май 2015 года. Данные опубликованы в открытом доступе на платформе Kaggle. 

Данные с сайта renthub.com по стоимости квартир для аренды:

<a href="https://drive.google.com/uc?id=1Dr-p-o5yMi3uIS2kPcUUwT5bH0cCt1Be
" target="_blank"><img src="https://drive.google.com/uc?id=1Dr-p-o5yMi3uIS2kPcUUwT5bH0cCt1Be" 
alt="IMAGE ALT TEXT HERE" width="360" border="0" /></a>


**Обзор доступных данных**

В наборе данных 21613 наблюдений и 16 признаков. 
Таким образом, про каждый из 21613 объектов недвижимости мы знаем значения 16 характеристик (число спален, оценка состояния риелтором, наличие вида на воду и т.п.)

Выборка была разбита на две части для обучения и для тестирования модели. Данные на обучение и тест обычно делят случайно и равномерно: нужно чтобы и обучающая, и тестовая выборка были похожие на те данные, на которых модель будет использоваться в боевых условиях. 

Более сложные разбиения могут использоваться если в данных есть временная компонента - например, мы строим скоринговую модель и для теста откладываем данные за последние несколько месяцев, а всю остальную выборку используем для обучения. Такое разбиение имитирует реальную работу модели: мы можем обучаться только на данных прошлых кредитов, а работать модель уже будет на данных из будущего. 

Доли общей выборки для обучения и тестирования обычно 70% и 30% соотвественно. Любые разумные числа подходят, если для обучения используется достаточно много данных (обычно больше 50%), но и для тестирования что-то остается (10% и больше).


Данные содержат два типа переменных:

* Целевая: **Целевая. Цена**
* Остальные переменные **могут использоваться для прогноза целевой переменной.**

## План анализа данных (data mining):

  1. Загрузить данные для обучения
  2. Обработать данные перед обучением модели
  3. Обучить модель на обучающей выборке
  4. Загрузить и предобработать данные для тестирования
  5. Провалидировать модель на тестовой выборке
  6. Проанализировать важность признаков и не забыть про постановку задачи

## 1. Загрузить данные для обучения

**Шаг 1.1. Загружаем библиотеки** 

Библиотека **warnings** отвечает за то, какие предупреждения (warnings) о работе будут выводиться пользователю. 
FutureWarning - предупреждения о том, как изменится работа библиотек в будущих версиях.
Поэтому такие предупреждения мы будем игнорировать.
Чтобы включить режим игнорирования мы отбираем все предупреждения из категории FutureWarning и выбираем для них действия 'ignore'.
Это делается вызовом функции simplefilter c задание двух атрибутов: действия action и категории предупреждений category.

In [ ]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

Для корректной работы с данными в python требуется загрузить специальную библиотеку
**pandas**, программную библиотеку на языке python для обработки и анализа данных. 

In [ ]:
import pandas as pd # загружаем библиотеку и для простоты обращения в коде называем её сокращенно pd

Для корректной работы с графиками в python требуется загрузить специальную библиотеку
**matplotlib**, программную библиотеку на языке python для визуализации данных двумерной и трехмерной графикой.

Графики используются для облегчения интерпретации полученных результатов, а также в качестве иллюстраций в презентациях и отчетах. 

Оснвные методы для построения:
* plot() - графики
* semilogy() - график логарифмический
* hist() - гистограммы

In [ ]:
import matplotlib.pyplot as plt # загружаем библиотеку и для простоты обращения в коде называем её сокращенно plt
# указываем, чтобы картинки отображались прямо в ноутбуке 
%matplotlib inline 

**Шаг 1.2. Загрузим данные**

Для решения задачи мы будем использовать данные. Они состоят из двух частей: часть для обучения и часть для тестирования модели. Загружаем данные с помощие команды !wget. Для того, чтобы игнорировать сообщения в процессе загрузки используем магическую команду %%capture в первой строке.

In [ ]:
%%capture
!wget https://www.dropbox.com/s/afwb0tnqm9izxha/predict_house_price_training_data.xlsx
!wget https://www.dropbox.com/s/sur2avqf4n5f4az/predict_house_price_test_data.xlsx

Так как данные в формате xlsx (Excel), мы будем использовать специальную функцию
из библиотеки pandas для загрузки таких данных **read_excel**.

В функции передаем один атрибут: название таблицы с данными.

In [ ]:
training_data = pd.read_excel('predict_house_price_training_data.xlsx') # загружаем таблицу в переменную training_data

*Что важно посмотреть после того, как мы загрузили данные?*
- проверить, что данные действительно загрузились
- посмотреть на данные, чтобы удостовериться, что они правильные: колонки имеют те же названия, что и в таблице и т.д.

Для того чтобы это сделать, нужно вызвать от переменной *training_data* метод **head()**, который выводит первые 5 строк таблицы. 

Для вызова метода объекта необходимо сначала написать *имя объекта*, затем поставить *точку*, затем уже написать *название метода*. Обратите внимание, что в конце обязательно ставить скобочки, потому что метод - это функция и в ней есть аргументы, просто в данном случае мы их не передаем, поэтому оставляем поле пустым

In [ ]:
training_data.head()

Целевая.Цена  Спальни  Ванные  Жилая площадь  Общая площадь  \
0        830000        5    3.50           3490          21780   
1        385000        4    1.75           2360           7620   
2        610000        6    2.75           2040           8560   
3        550000        3    1.75           1940           8376   
4       1300000        3    2.75           3450           5350   

   Количество этажей  Вид на воду  Просмотрены ранее  Состояние  \
0                2.0            0                  0          3   
1                1.0            0                  0          4   
2                1.0            0                  2          4   
3                1.0            0                  0          4   
4                1.5            0                  3          4   

   Оценка риелтора  Площадь без подвала  Площадь подвала  Год постройки  \
0                8                 3490                0           1996   
1                7                 1180             1180           1955   
2                7                 1100              940           1961   
3                8                 1290              650           1963   
4                9                 2590              860           1925   

   Год реновации   Широта  Долгота  
0              0  47.6707 -122.144  
1              0  47.5278 -122.345  
2              0  47.6160 -122.115  
3              0  47.5586 -122.173  
4              0  47.6389 -122.407

**Шаг 1.3. Посмотрим на размеры загруженной таблицы**, у которой мы видели только первые 5 строк.

Для этого вызываем поле **shape** у нашей переменной *training_data*. Поле вызывается также как метод, но в конце скобки не ставятся, так как для поля не предусмотрена передача аргументов.  

In [ ]:
training_data.shape

(15129, 16)

*Что означает первое и второе число?*

Итак, таблица содержит 15129 строк (объектов) и 16 столбцов (признаков), включая выходной (целевой) признак.
15129 меньше 21613, так как мы пока загрузили только часть данных, которую будем использовать для обучения модели.

Таблицу проверили, теперь можно приступать к обработке данных.

# 2. Обработать данные перед обучением модели

**Шаг 2.1. Проверяем данные на наличие пропусков и типов переменных**

Начнем с проверки общей информации о данных.
Для того чтобы это сделать, нужно обратиться вызвать у переменной *training_data* метод **info()**.

Напомним, что в конце необходимо поставить скобочки.

In [ ]:
training_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15129 entries, 0 to 15128
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Целевая.Цена         15129 non-null  int64  
 1   Спальни              15129 non-null  int64  
 2   Ванные               15129 non-null  float64
 3   Жилая площадь        15129 non-null  int64  
 4   Общая площадь        15129 non-null  int64  
 5   Количество этажей    15129 non-null  float64
 6   Вид на воду          15129 non-null  int64  
 7   Просмотрены ранее    15129 non-null  int64  
 8   Состояние            15129 non-null  int64  
 9   Оценка риелтора      15129 non-null  int64  
 10  Площадь без подвала  15129 non-null  int64  
 11  Площадь подвала      15129 non-null  int64  
 12  Год постройки        15129 non-null  int64  
 13  Год реновации        15129 non-null  int64  
 14  Широта               15129 non-null  float64
 15  Долгота              15129 non-null 

Анализируем результата выполнения команды:

* 15129 строк (entries)
* 16 столбцов (Data columns)

В данных присутствует всего два типа dtypes:
* int64 - целое число  (12 столбцов)
* float64 - дробное число (4 столбца)

Цифры в каждой строчке обозначают количество заполненных (*non-null*) значений. Так как эти цифры в каждой строчке совпадают с числом строк (15129), то в данных нет пропусков и можно двигаться дальше. 

**Шаг 2.2. Работаем с целевой переменной**

*Какая переменная целевая?*

В данном случае по условию задачи мы должны прогнозировать стоимость, поэтому целевая переменная - это цена. 

In [ ]:
target_variable_name = 'Целевая.Цена'

Нам нужно выделить в отдельную переменную *training_values* столбец из нашей таблицы, который соответствует определенной выше целевой переменной. Для этого мы у таблицы *training_data* в квадратных скобках указываем имя нужного столбца. В нашем случае это имя записано в переменной *target_variable_name*. 

In [ ]:
training_values = training_data[target_variable_name]

Отделим входные переменные от выходной (целевой), чтобы можно было построить модель предсказания целевой переменной по входным. 
Для это нужно у переменной *training_data* вызвать метод **drop()**. Результат мы записываем в новую переменную *training_points*. После выполнения запроса *training_points* будет содержать исходную таблицу без целевого столбца. 

Обратите внимание, что в данном случае мы передаем два аргумента:
    1. target_variable_name - название столбца цены, который мы ранее записали в эту переменную и теперь хотим удалить из training_data
    2. axis=1 - означает, что мы удаляем столбец, а в случае axis=0 - означает, что мы удаляем строку

In [ ]:
training_points = training_data.drop(target_variable_name, axis=1)

Можно посмотреть результаты этих действий, вызвав метод **head()** и поле **shape**, которыми мы пользовались ранее, но сейчас нужно вызывать их от новой переменной *training_points*.

In [ ]:
training_points.head()

Спальни  Ванные  Жилая площадь  Общая площадь  Количество этажей  \
0        5    3.50           3490          21780                2.0   
1        4    1.75           2360           7620                1.0   
2        6    2.75           2040           8560                1.0   
3        3    1.75           1940           8376                1.0   
4        3    2.75           3450           5350                1.5   

   Вид на воду  Просмотрены ранее  Состояние  Оценка риелтора  \
0            0                  0          3                8   
1            0                  0          4                7   
2            0                  2          4                7   
3            0                  0          4                8   
4            0                  3          4                9   

   Площадь без подвала  Площадь подвала  Год постройки  Год реновации  \
0                 3490                0           1996              0   
1                 1180             1180           1955              0   
2                 1100              940           1961              0   
3                 1290              650           1963              0   
4                 2590              860           1925              0   

    Широта  Долгота  
0  47.6707 -122.144  
1  47.5278 -122.345  
2  47.6160 -122.115  
3  47.5586 -122.173  
4  47.6389 -122.407

In [ ]:
training_points.shape

(15129, 15)

Видно, что столбца действительно нет, а количество строк не изменилось. Данные в 5 первых строках такие же, как были ранее.

##   3. Обучить модель на обучающей выборке

**Шаг 3.1. Выбираем метод, который будем использовать**

Проще всего начать с простых методов. 

На выбор метода для построения модели влияет набор признаков, размер выборки, интуиция про то, какая связь между входными переменными и целевой. Но часто решение принимается исходя из того, какая модель сработала лучше.

Для корректной работы с методами построения моделей в python требуется загрузить специальную библиотеку
**sklearn**, программную библиотеку на языке python для для машинного обучения и анализа данных.

Импортируем модуль из этой библиотеки:
 * *linear_model* - тут находятся все линейные модели

In [ ]:
from sklearn import linear_model

Чтобы создать модель линейной регресии, пишем имя модуля 'linear_model', затем точку, затем название модели.

In [ ]:
linear_regression_model = linear_model.LinearRegression() # создаем модель

In [ ]:
linear_regression_model # смотрим, что получилось

LinearRegression()

**Шаг 3.2. Обучить модель**

Теперь, когда мы создали прототип модели, можем её обучить на обучающей выборке. 

Для этого вызываем метод **fit()** и передаем ему на вход два аргумента: 
таблицу входных признаков и столбец значений целевой переменной - (training_points, training_values)

In [ ]:
linear_regression_model.fit(training_points, training_values)

LinearRegression()

## Задание 3.5

Реализуйте настройку параметров модели не с помощью встроенной функции, а с помощью своей реализации.

Настроить параметры нужно либо с помощью **стохастического градиентного спуска** либо с помощью **mini-batch градиентного спуска**. 
Нужно использовать регуляризацию **L1, L2 или Elastic Net**. 

In [ ]:
import numpy as np
from sklearn.base import BaseEstimator, RegressorMixin
from numpy.linalg import inv

class NewLinearRegression(BaseEstimator, RegressorMixin):
  def __init__(self, elastic_net = False, elastic_net_coefs=[0., 0.]):
    self.elastic_net = elastic_net
    self.elastic_net_coefs = [float(elastic_net_coefs[0]), float(elastic_net_coefs[1])]
  
  def _step(self, X, y, learning_rate):
    if not self.elastic_net:
      self.w -= 2 * learning_rate * np.dot(X.T, np.dot(X, self.w) - y)
    else:
      self.w -= 2 * learning_rate * (np.dot(X.T, np.dot(X, self.w) - y) + self.elastic_net_coefs[0] * self.w + 
                                     self.elastic_net_coefs[1] * ((self.w > 0).astype(float) - (self.w < 0).astype(float)))


  def fit(self, X, y, learning_rate = 10e-4, epochs = 10, eps_limit = 10e-4, learning_rate_impact_per_epoch = 1):
    X = np.column_stack((X, np.ones(X.shape[0]).astype(float)))
    self.w = (np.random.uniform(-1, 1, X.shape[1]) - 0.5).copy()
    for epoch in range(epochs):
      learning_rate = learning_rate / learning_rate_impact_per_epoch
      for example in range(y.size):
        self._step(X[example], y[example], learning_rate)
    return self

  def predict(self, X):
    X = np.column_stack((X, np.ones(X.shape[0]).astype(float)))
    return X @ self.w

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error

scaler = StandardScaler().fit(training_points)
training_points_scaled = scaler.transform(training_points)
# print(training_points_scaled)

lr = NewLinearRegression().fit(training_points_scaled, training_values)
y_pred = lr.predict(training_points_scaled)
print(len(training_points_scaled))
print('MAE:', mean_absolute_error(y_pred, training_values))
print('MSE:', mean_squared_error(y_pred, training_values))
print(lr.w)
print(y_pred)

15129
MAE: 124757.54234875273
MSE: 42687275632.75769
[-39983.08310977  35711.45107149  83518.45763508  -3877.49475921
  -4763.6006714   27877.63175911  36999.00648142  24078.83011476
 108983.54496899  75527.58236705  31335.40779947 -60758.35510211
   1932.35122584  81207.69934496 -21618.95351827 536185.3704422 ]
[795650.77268534 512203.47505789 519696.74462919 ... 762476.74081424
 488402.15512993 495161.64105677]


# 4. Загрузить и предобработать данные для тестирования

**Шаг 4.1. Загрузим и проанализируем тестовые данные.**

Так как данные в формате xlsx (Excel), мы будем использовать специальную функцию
из библиотеки pandas для загрузки таких данных **read_excel**.

В функции передаем один атрибут: название файла, в котором находится таблица с данными.

In [ ]:
test_data = pd.read_excel('predict_house_price_test_data.xlsx')

*Что важно посмотреть, после того, как мы загрузили данные?*

- проверить, что данные действительно загрузились
- посмотреть на данные, чтобы удостовериться, что они правильные: колонки имеют те же названия, что и в таблице и т.д.

Для того чтобы это сделать, нужно вызвать от переменной *test_data* метод **head()**, который выводит первые 5 строк таблицы. 

Для вызова метода объекта необходимо сначала написать *имя объекта*, затем поставить *точку*, затем уже написать *название метода*. Обратите внимание, что в конце обязательно ставить скобочки, потому что метод - это функция и в ней есть аргументы, просто в данном случае мы их не передаем, поэтому оставляем поле пустым

In [ ]:
test_data.head()

Целевая.Цена  Спальни  Ванные  Жилая площадь  Общая площадь  \
0        260000        3    1.00           1300          10139   
1        734500        4    2.75           3280           6845   
2        325000        1    1.00           1220          12426   
3       1990000        3    2.50           2880          13500   
4        315000        3    2.00           1300           3731   

   Количество этажей  Вид на воду  Просмотрены ранее  Состояние  \
0                1.0            0                  0          3   
1                2.0            0                  0          3   
2                1.0            0                  4          4   
3                1.0            0                  4          5   
4                1.0            0                  0          3   

   Оценка риелтора  Площадь без подвала  Площадь подвала  Год постройки  \
0                7                 1300                0           1962   
1               10                 3280                0           2003   
2                6                 1220                0           1946   
3                8                 1520             1360           1950   
4                7                  900              400           1993   

   Год реновации   Широта  Долгота  
0           2007  47.3427 -122.087  
1              0  47.7042 -122.107  
2              0  47.4047 -122.331  
3              0  47.6281 -122.216  
4              0  47.5374 -122.270

Посмотрим на размеры загруженной таблицы, так как мы видели только 5 строк

Для этого вызываем поле **shape** у нашей переменной *test_data*. Поле вызывается также как метод, но в конце скобки не ставятся (!), так как для поля не предусмотрена передача аргументов.  

In [ ]:
test_data.shape

(6484, 16)

*Что означает первое и второе число?* 
Таблица содержит 6484 строк (объектов) и 16 столбцов (признаков), включая выходной (целевой) признак. Также как в учебных данных до обучения.

Таблицу проверили, теперь можно приступать к обработке данных
Действуем аналогично тому, как делали с данными для обучения

Проверим, есть ли в данных пропуски. Для того чтобы это сделать, нужно обратиться вызвать у переменной *test_data* метод **info()**.

In [ ]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6484 entries, 0 to 6483
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Целевая.Цена         6484 non-null   int64  
 1   Спальни              6484 non-null   int64  
 2   Ванные               6484 non-null   float64
 3   Жилая площадь        6484 non-null   int64  
 4   Общая площадь        6484 non-null   int64  
 5   Количество этажей    6484 non-null   float64
 6   Вид на воду          6484 non-null   int64  
 7   Просмотрены ранее    6484 non-null   int64  
 8   Состояние            6484 non-null   int64  
 9   Оценка риелтора      6484 non-null   int64  
 10  Площадь без подвала  6484 non-null   int64  
 11  Площадь подвала      6484 non-null   int64  
 12  Год постройки        6484 non-null   int64  
 13  Год реновации        6484 non-null   int64  
 14  Широта               6484 non-null   float64
 15  Долгота              6484 non-null   f

Цифры в каждой строчке обозначают количество заполненных (*non-null*) значений. Так как эти цифры в каждой строчке совпадают с числом строк (6484), то в данных нет пропусков.

**Шаг 4.2. Отделяем целевую переменную**

Нам нужно выделить в отдельную переменную *test_values* столбец из нашей таблицы, который соответствует определенной выше целевой переменной. Для этого мы у таблицы *test_data* в квадратных скобках указываем имя нужного столбца. В нашем случае это имя записано в переменной *target_variable_name*. 

In [ ]:
test_values = test_data[target_variable_name]

Отделим входные переменные от выходной (целевой), чтобы можно было построить модель предсказания целевой переменной по входным. 
Для это нужно у переменной *test_data* вызвать метод **drop()**. Результат мы записываем в новую переменную *test_points*. После выполнения запроса *test_points* будет содержать исходную таблицу без целевого столбца. 

Обратите внимание, что в данном случае мы передаем два аргумента:
    1. target_variable_name - название столбца цены, который мы ранее записали в эту переменную и теперь хотим удалить из training_data
    2. axis=1 - означает, что мы удаляем столбец, а в случае axis=0 - означает, что мы удаляем строку

In [ ]:
test_points = test_data.drop(target_variable_name, axis=1)

И проверяем результат записанный в test_points

In [ ]:
test_points.head()

Спальни  Ванные  Жилая площадь  Общая площадь  Количество этажей  \
0        3    1.00           1300          10139                1.0   
1        4    2.75           3280           6845                2.0   
2        1    1.00           1220          12426                1.0   
3        3    2.50           2880          13500                1.0   
4        3    2.00           1300           3731                1.0   

   Вид на воду  Просмотрены ранее  Состояние  Оценка риелтора  \
0            0                  0          3                7   
1            0                  0          3               10   
2            0                  4          4                6   
3            0                  4          5                8   
4            0                  0          3                7   

   Площадь без подвала  Площадь подвала  Год постройки  Год реновации  \
0                 1300                0           1962           2007   
1                 3280                0           2003              0   
2                 1220                0           1946              0   
3                 1520             1360           1950              0   
4                  900              400           1993              0   

    Широта  Долгота  
0  47.3427 -122.087  
1  47.7042 -122.107  
2  47.4047 -122.331  
3  47.6281 -122.216  
4  47.5374 -122.270

In [ ]:
test_points.shape

(6484, 15)

Видно, что количество столбцов уменьшилось на один. Данные в 5 первых строках такие же, как были ранее.

# 5. Провалидировать модель на тестовой выборке

**Шаг 5.1. Сравнение моделей.**

Теперь мы готовы сравнить качество двух моделей! 😎

## Задание 3.6 Какая модель лучше?

Получим прогнозы целевой переменной на тестовых данных для модели линейной регрессии из sklearn и для своей реализации. 



In [ ]:
test_predictions_linear = linear_regression_model.predict(test_points)

In [ ]:
### введите свой код
### код для получения предсказаний с помощью своей реализации алгоритма настройки параметров модели

test_points_scaled = scaler.transform(test_points)
lr_pred = lr.predict(test_points_scaled)
print(test_points_scaled.shape)
print(training_values.shape)
print(training_points_scaled.shape)
print(lr.w.shape)


print(test_points_scaled)
print(lr_pred)

(6484, 15)
(15129,)
(15129, 15)
(16,)
[[-0.401407   -1.45015324 -0.8522375  ...  4.78855538 -1.57047897
   0.90281664]
 [ 0.67021581  0.83271773  1.32115491 ... -0.20999201  1.03972929
   0.76132176]
 [-2.54465261 -1.45015324 -0.94005134 ... -0.20999201 -1.12280839
  -0.82342089]
 ...
 [-0.401407    1.811091    1.59557314 ... -0.20999201  0.90615017
  -0.88001884]
 [ 0.67021581  0.18046888  0.21250524 ... -0.20999201 -1.69033915
  -0.99321474]
 [-0.401407    0.18046888  0.344226   ... -0.20999201  0.65920931
  -0.44845946]]
[ 159606.05957101  952354.82210245  464150.03909105 ... 1067225.04631604
  325634.64170128  966749.40862534]


Качество регрессионных моделей оценим двумя способами: 
1. Сравним визуально прогнозы с настоящими ценами (тестовые с предсказанием)
2. Сравним метрики качества MSE и MAE

Для корректного подсчета метрик качества модели в python требуется загрузить их из библиотеки **sklearn**. 

Мы используем две метрики качества:
 * *mean_absolute_error* - средняя абсолютная ошибка $|y_i - \hat{y}_i|$
 * *mean_squared_error* - средняя квадратичная ошибка $(y_i - \hat{y}_i)^2$

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

Подсчитаем ошибки для линейной модели.

Для этого вызовем методы **mean_absolute_error()** и **mean_squared_error()**. На вход им передается столбец настоящих значений *test_values* и столбец значений, предсказанных моделью линейной регрессии *test_predictions_linear*.

In [ ]:
mean_absolute_error_linear_model = mean_absolute_error(test_values, test_predictions_linear) 
mean_squared_error_linear_model = mean_squared_error(test_values, test_predictions_linear)

Подсчитайте ошибки для своей реализации модели.

In [ ]:
# mse mae
mae_model = mean_absolute_error(test_values, lr_pred) 
mse_model = mean_squared_error(test_values, lr_pred)

Теперь напечатаем полученные ошибки. Обычно смотрят на корень из среднеквадратичной ошибки, RMSE. Чтобы извлечь корень нам понадобится библиотека **Numpy**. С помощью неё можно быстро производить вычисления сразу над массивами чисел.

In [ ]:
print('sklearn:\nMAE:', mean_absolute_error_linear_model)
print('MSE:', mean_squared_error_linear_model)
print('RMSE:', np.sqrt(mean_squared_error_linear_model))

print('\nmodel:\nMAE:', mae_model)
print('MSE:', mse_model)
print('RMSE:', np.sqrt(mse_model))


sklearn:
MAE: 126852.51255005135
MSE: 40756843765.100815
RMSE: 201883.24290317117

model:
MAE: 125238.67302458607
MSE: 41722710550.296005
RMSE: 204261.37801918405


## 6. Выявление важных признаков

Признаки, используемые для обучения модели, оказывают большое влияние на качество результатов. Неинформативные или слабо информативные признаки могут снизить качество модели, использование избыточных признаков в модели замедляет скорость ее работы.

Для того, чтобы отобрать признаки нам следует ответить на следующие вопросы: 

    1. Все ли признаки в наших данных заполненны разумными значениями?
    2. Какие признаки будут больше всего влиять на значение целевой переменной?
    3. Какие дополнительные признаки имело бы смысл добавить в список входных?

**6.1. Разглядывание значений признаков**

Для того чтобы ответить на первый вопрос, нам необходимо получить некоторую нашу экспертную оценку данных. В наших данных нет пропусков, однако иногда пропуски предварительно заполняют другими значениями (0, 999 и другие) и их можно увидеть, только внимательно посмотрев на данные. 

Для того, чтобы это сделать, нужно вызвать у переменной *training_points* метод **head(10)**, который выводит первые 10 строк таблицы.

In [ ]:
training_points.head(10)

Спальни  Ванные  Жилая площадь  Общая площадь  Количество этажей  \
0        5    3.50           3490          21780                2.0   
1        4    1.75           2360           7620                1.0   
2        6    2.75           2040           8560                1.0   
3        3    1.75           1940           8376                1.0   
4        3    2.75           3450           5350                1.5   
5        3    2.25           2300           9914                2.0   
6        3    2.50           2770           8820                1.0   
7        2    1.75           1650           7500                1.0   
8        1    1.00            580           1799                1.0   
9        2    1.00            900           3400                1.0   

   Вид на воду  Просмотрены ранее  Состояние  Оценка риелтора  \
0            0                  0          3                8   
1            0                  0          4                7   
2            0                  2          4                7   
3            0                  0          4                8   
4            0                  3          4                9   
5            0                  0          4                8   
6            0                  0          3                7   
7            0                  0          4                7   
8            0                  0          3                7   
9            0                  0          5                6   

   Площадь без подвала  Площадь подвала  Год постройки  Год реновации  \
0                 3490                0           1996              0   
1                 1180             1180           1955              0   
2                 1100              940           1961              0   
3                 1290              650           1963              0   
4                 2590              860           1925              0   
5                 2300                0           1980              0   
6                 1900              870           1980           2004   
7                 1000              650           1959              0   
8                  580                0           1908           2005   
9                  900                0           1905              0   

    Широта  Долгота  
0  47.6707 -122.144  
1  47.5278 -122.345  
2  47.6160 -122.115  
3  47.5586 -122.173  
4  47.6389 -122.407  
5  47.5677 -122.086  
6  47.3685 -122.048  
7  47.6871 -122.207  
8  47.6829 -122.375  
9  47.5269 -122.314

Рассмотрим столбец "Год реновации". Он содержит много нулей и мало реальных данных о годах, что как раз свидетельствует о том, что в этих данных изначально были пропуски, а к нам они попали уже заполненные нулями. 

Мы можем посмотреть количество уникальных значений и сколько раз эти значения встречаются в этом столбце. Для этого вызываем метод **value_counts()** у нашего столбца *training_points['Год реновации']*

In [ ]:
training_points['Год реновации'].value_counts()

0       14490
2014       63
2013       31
2000       28
2003       24
        ...  
1948        1
1950        1
1953        1
1934        1
1955        1
Name: Год реновации, Length: 67, dtype: int64

Как видно, нулей существенно больше (14490), чем любых других значений (639), поэтому можем сделать вывод, что этот признак не информативен и может негативно повлиять на качество модели.

**6.2. Какие признаки самые важные**

Ручной перебор признаков невозможен, если их тысячи, кроме того, часто даже эксперты не могут сказать, использование каких признаков в модели даст прирост качества.
Кроме ручного перебора всех признаков и их визуального анализа, можно использовать оценку значимости признаков с помощью обученных моделей. 

## Задание 3.7 

In [ ]:
## введите свой код
## напишите код, который вычисляет значимость признака (фактически это просто вес по модулю)

# pd.DataFrame(abs(lr.w), test_points.columns)
# print(lr.w)
coeffs = lr.w
# print(coeffs)
coeffs = coeffs.tolist()
coeffs.pop()
coeffs = np.array(coeffs)
# print(coeffs)
pd.DataFrame(abs(coeffs), test_points.columns)

0
Спальни               39983.083110
Ванные                35711.451071
Жилая площадь         83518.329790
Общая площадь          3877.494759
Количество этажей      4763.600671
Вид на воду           27877.631759
Просмотрены ранее     36999.006481
Состояние             24078.830115
Оценка риелтора      108983.544969
Площадь без подвала   75527.698189
Площадь подвала       31335.469376
Год постройки         60758.355102
Год реновации          1932.351226
Широта                81207.699345
Долгота               21618.953518

Представьте полученные результаты в более удобном табличном виде. Создайте пустую таблицу *feature_importance* с двумя колонками "Название признака", "Важность признака". Для этого воспользуемся функцией **DataFrame()** из библиотеки pandas и передадим ей в качестве аргумента   
*columns = ["Название признака", "Важность признака"]*

In [ ]:
feature_importance = pd.DataFrame(columns = ['Название признака', 'Важность признака'])
feature_importance['Название признака'] = training_points.columns
feature_importance['Важность признака'] = np.abs(coeffs)

Заполните соответствующие колонки значениями, которые мы получили выше. 

In [ ]:
# выведите полученную таблицу
feature_importance

Название признака  Важность признака
0               Спальни       39983.083110
1                Ванные       35711.451071
2         Жилая площадь       83518.329790
3         Общая площадь        3877.494759
4     Количество этажей        4763.600671
5           Вид на воду       27877.631759
6     Просмотрены ранее       36999.006481
7             Состояние       24078.830115
8       Оценка риелтора      108983.544969
9   Площадь без подвала       75527.698189
10      Площадь подвала       31335.469376
11        Год постройки       60758.355102
12        Год реновации        1932.351226
13               Широта       81207.699345
14              Долгота       21618.953518

Мы можем отсортировать значения по важности, вызвав у *feature_importance* метод **sort_values()**.

Обратите внимание, что в данном случае мы передаем два аргумента:

1. by='Важность признака' - название столбца, по которому мы сортируем
2. ascending=False - означает, что мы сортируем по убиыванию, а в случае ascending=True сортировка будет по возрастанию

In [ ]:
feature_importance.sort_values(by='Важность признака', ascending=False)

Название признака  Важность признака
8       Оценка риелтора      108983.544969
2         Жилая площадь       83518.329790
13               Широта       81207.699345
9   Площадь без подвала       75527.698189
11        Год постройки       60758.355102
0               Спальни       39983.083110
6     Просмотрены ранее       36999.006481
1                Ванные       35711.451071
10      Площадь подвала       31335.469376
5           Вид на воду       27877.631759
7             Состояние       24078.830115
14              Долгота       21618.953518
4     Количество этажей        4763.600671
3         Общая площадь        3877.494759
12        Год реновации        1932.351226

## Задание 3.8. 

Напомню исходную постановку задачи. Мы говорили о том, что оценка цены на недвижимость может быть нужна для выявления аномально низких цен на объект недвижимости. Давайте попробуем найти такие в датасете.

In [ ]:
average_deviation = mean_absolute_error_linear_model / training_points.shape[0]
test_data['diff'] = test_values - linear_regression_model.predict(test_points)

less_av_mask = test_data['diff'] < 0
test_data[less_av_mask].sort_values(by='diff', ascending=True).head(20)

Целевая.Цена  Спальни  Ванные  Жилая площадь  Общая площадь  \
632         425000        3    2.75           3610         107386   
4675        380000        3    2.50           1980          17342   
1032       2280000        7    8.00          13540         307752   
4270        995000        3    4.50           4380          47044   
3502        400000        3    2.50           2090          32718   
2765        357000        3    2.00           2460          53882   
2426        600000        2    2.25           2320          18919   
4838       1110000        5    3.50           7350          12231   
842         345600        5    3.50           2800           5120   
4643        650000        3    1.75           2800          19386   
3511        600000        2    1.75           1550           7764   
4354        600000        3    2.50           4930          77536   
2739        603500        4    2.50           4060           9734   
6303        452000        5    2.50           5067          13315   
277         500000        2    1.00           1760          27332   
4438        650000        5    3.50           5110          10018   
1834        540000        7    5.75           3700           7647   
6088        750000        4    1.75           2520          21834   
4178       1000000        5    4.50           2120           8944   
6180        480000        2    1.50           1008          26487   

      Количество этажей  Вид на воду  Просмотрены ранее  Состояние  \
632                 1.5            1                  3          3   
4675                2.0            1                  4          3   
1032                3.0            0                  4          3   
4270                2.0            1                  3          3   
3502                2.0            1                  4          3   
2765                1.0            1                  4          3   
2426                2.0            1                  4          4   
4838                2.0            0                  4          3   
842                 2.5            0                  0          3   
4643                1.0            1                  4          3   
3511                1.0            1                  4          4   
4354                2.0            0                  0          3   
2739                1.0            0                  4          3   
6303                1.0            0                  2          3   
277                 1.0            1                  4          4   
4438                2.0            0                  0          3   
1834                2.0            0                  1          3   
6088                1.0            1                  4          3   
4178                2.0            1                  4          5   
6180                1.0            1                  4          4   

      Оценка риелтора  Площадь без подвала  Площадь подвала  Год постройки  \
632                 8                 3130              480           1918   
4675               10                 1580              400           1984   
1032               12                 9410             4130           1999   
4270                9                 3720              660           1968   
3502                7                 1550              540           1919   
2765                7                 2460                0           1955   
2426                8                 2320                0           1976   
4838               11                 4750             2600           2001   
842                 9                 2800                0           1903   
4643                8                 1400             1400           1965   
3511                8                 1550                0           1965   
4354                9                 3930             1000           1981   
2739                9                 2150             1910           1